File: leetcode/lru_cache/LRUCache.cpp
#include <iostream> 
#include <map> 

using namespace std; 

typedef struct CacheNode {
	int key; 
    int val; 
    CacheNode * next; 
    CacheNode * prev; 
    CacheNode(int n, int k): val(n), key(k), next(NULL), prev(NULL){}
} CacheNode; 
    
class LRUCache{
private: 
    map<int, CacheNode*> cacheMap; 
    CacheNode * head; 
    CacheNode * tail; 
    int cap; 
    int size; 
public:
    LRUCache(int capacity) {
        cap = capacity; 
        head = NULL; 
        tail = NULL; 
        size = 0; 
    }
    
    int get(int key) {
        if (cacheMap.find(key) == cacheMap.end()) {
            return -1; 
        }
        
        CacheNode * node = cacheMap[key]; 
        int result = node->val; 
        
        if (node != head) {
            node->prev->next = node->next; 
            
            if (node->next != NULL) {
                node->next->prev = node->prev; 
            }
            else {  // At tail
                tail = node->prev; 
            }
            
            node->prev = NULL; 
            node->next = head; 
            head->prev = node; 
            head = node; 
        }
        
        return result; 
    }
    
    void set(int key, int value) {
        if (cacheMap.find(key) == cacheMap.end()) {  // New Element 
            CacheNode * newNode = new CacheNode(value, key); 
            cacheMap[key] = newNode;
            
            if (head == NULL) {
                head = newNode; 
                tail = newNode; 
            }
            else {
                newNode->next = head; 
                head->prev = newNode; 
                head = newNode; 
            } 
            
            if (size == cap) {  // Exceeds capacity, remove last one
                if (tail != NULL) { 
                	// Erase 
                	cacheMap.erase(tail->key); 
                    tail = tail->prev; 
                    if (tail != NULL) tail->next = NULL;  
                }
            }
            else {
            	size++; 
            }
        }
        else {  // Element already exist 
            CacheNode * node = cacheMap[key];
            node->val = value;  
            
            if (node != head) {
                node->prev->next = node->next; 
                
                if (node->next != NULL) {
                    node->next->prev = node->prev; 
                }
                else {  // At tail
                    tail = node->prev; 
                }
                
                node->prev = NULL; 
                node->next = head; 
                head->prev = node; 
                head = node; 
            }            
        }
    }
};

int main() {
	LRUCache cache(1); 
cache.set(2,1);
cout << cache.get(2) << endl;
cache.set(3,2);
cout << cache.get(2) << endl;
cout << cache.get(3) << endl;

	return 0; 
}

File: leetcode2/lru_cache/LRUCache.java
public class LRUCache {
    class Node {
        int key; 
        int val = 0; 
        Node next = null; 
        Node prev = null; 
        Node(int key, int val) {
            this.key = key; 
            this.val = val; 
        }
    }
    
    private int max = 0;
    private int cnt = 0; 
    private Map<Integer, Node> m = new HashMap<Integer, Node>(); 
    private Node head = null; 
    private Node tail = null; 
    
    public LRUCache(int capacity) {
        this.max = capacity; 
    }
    
    public int get(int key) {
        if (!m.containsKey(key)) return -1; 
        
        Node n = m.get(key); 
        int result = n.val; 
        
        if (n != head) {
            n.prev.next = n.next; 
            
            if (n.next != null) {
                n.next.prev = n.prev; 
            }
            else {  // Move tail 
                tail = n.prev; 
            }
            
            n.prev = null; 
            n.next = head; 
            head.prev = n; 
            head = n; 
        }

        return result; 
    }
    
    public void set(int key, int value) {
        if (m.containsKey(key)) {
            Node n = m.get(key); 
            n.val = value; 
            
            if (n != head) {
                n.prev.next = n.next; 
                
                if (n.next != null) {
                    n.next.prev = n.prev; 
                }
                else {
                    tail = n.prev; 
                }
                
                n.next = head; 
                n.prev = null; 
                head.prev = n; 
                head = n; 
            }
        }
        else {
            Node n = new Node(key, value); 
            m.put(key, n); 
            
            if (head != null) {
                n.next = head; 
                head.prev = n; 
                head = n; 
            }
            else {
                head = n; 
                tail = n; 
            }
                
            cnt++; 
            
            if (cnt > max) {
                if (tail != null) {
                    m.remove(tail.key); 
                    tail = tail.prev; 
                    
                    if (tail != null) {
                        tail.next = null; 
                    }
                    
                    cnt--;
                }
            }
        }
    }
}
